File: flutter_project/lib/screens/color_screen.dart
import 'package:flutter/material.dart';
import '../models/clock_settings.dart';
import '../widgets/clock_display.dart';

/// ColorScreen
///
/// Educational Note:
/// - Provides both preset swatches and custom HSV color slider controls.
/// - In Flutter, Color(0xAARRGGBB) stores color as an unsigned 32-bit integer.
/// - HSVColor.fromAHSV() allows intuitive Hue (0-360), Saturation (0-1), and Value (0-1) adjustments.
class ColorScreen extends StatefulWidget {
  final DateTime currentTime;
  final ClockSettings settings;
  final Function(Color) onSelectColor;

  const ColorScreen({
    super.key,
    required this.currentTime,
    required this.settings,
    required this.onSelectColor,
  });

  @override
  State<ColorScreen> createState() => _ColorScreenState();
}

class _ColorScreenState extends State<ColorScreen> {
  static const List<Map<String, dynamic>> presetColors = [
    {'name': 'White', 'color': Color(0xFFFFFFFF)},
    {'name': 'Dark / Black', 'color': Color(0xFF333333)},
    {'name': 'Red', 'color': Color(0xFFFF3B30)},
    {'name': 'Blue', 'color': Color(0xFF007AFF)},
    {'name': 'Green', 'color': Color(0xFF34C759)},
    {'name': 'Yellow', 'color': Color(0xFFFFCC00)},
    {'name': 'Purple', 'color': Color(0xFFAF52DE)},
    {'name': 'Orange', 'color': Color(0xFFFF9500)},
    {'name': 'Cyan', 'color': Color(0xFF00E5FF)},
    {'name': 'Neon Green', 'color': Color(0xFF39FF14)},
    {'name': 'Hot Pink', 'color': Color(0xFFFF2D55)},
    {'name': 'Amber', 'color': Color(0xFFFFB300)},
  ];

  double _hue = 180.0;
  double _saturation = 1.0;
  double _value = 1.0;

  @override
  void initState() {
    super.initState();
    final hsv = HSVColor.fromColor(widget.settings.clockColor);
    _hue = hsv.hue;
    _saturation = hsv.saturation;
    _value = hsv.value;
  }

  void _updateCustomColor() {
    final color = HSVColor.fromAHSV(1.0, _hue, _saturation, _value).toColor();
    widget.onSelectColor(color);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFF0D1117),
      appBar: AppBar(
        title: const Text('Clock Color', style: TextStyle(fontWeight: FontWeight.bold)),
        backgroundColor: const Color(0xFF0D1117),
        elevation: 0,
      ),
      body: ListView(
        padding: const EdgeInsets.all(16.0),
        children: [
          // Live Clock Preview
          Container(
            padding: const EdgeInsets.all(24.0),
            decoration: BoxDecoration(
              color: Colors.black.withOpacity(0.4),
              borderRadius: BorderRadius.circular(20),
            ),
            child: Center(
              child: ClockDisplay(
                dateTime: widget.currentTime,
                settings: widget.settings,
                isFullScreen: false,
                previewScale: 0.85,
              ),
            ),
          ),
          const SizedBox(height: 20),

          const Text(
            'Preset Colors',
            style: TextStyle(color: Colors.white, fontSize: 16, fontWeight: FontWeight.bold),
          ),
          const SizedBox(height: 12),

          // Preset Swatches Grid
          Wrap(
            spacing: 12,
            runSpacing: 12,
            children: presetColors.map((item) {
              final color = item['color'] as Color;
              final isSelected = widget.settings.clockColor.value == color.value;

              return GestureDetector(
                onTap: () {
                  widget.onSelectColor(color);
                },
                child: Container(
                  width: 50,
                  height: 50,
                  decoration: BoxDecoration(
                    color: color,
                    shape: BoxShape.circle,
                    border: Border.all(
                      color: isSelected ? Colors.white : Colors.white24,
                      width: isSelected ? 3 : 1,
                    ),
                  ),
                  child: isSelected
                      ? const Icon(Icons.check, color: Colors.black, size: 24)
                      : null,
                ),
              );
            }).toList(),
          ),

          const SizedBox(height: 24),

          // Custom HSV Picker Section
          Container(
            padding: const EdgeInsets.all(18),
            decoration: BoxDecoration(
              color: Colors.white.withOpacity(0.05),
              borderRadius: BorderRadius.circular(20),
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    const Text(
                      'Custom Color Picker (HSV)',
                      style: TextStyle(color: Colors.white, fontSize: 15, fontWeight: FontWeight.bold),
                    ),
                    Container(
                      width: 24,
                      height: 24,
                      decoration: BoxDecoration(
                        color: widget.settings.clockColor,
                        shape: BoxShape.circle,
                        border: Border.all(color: Colors.white),
                      ),
                    ),
                  ],
                ),
                const SizedBox(height: 14),

                Text('Hue: ${_hue.toInt()}°', style: const TextStyle(color: Colors.white70, fontSize: 12)),
                Slider(
                  value: _hue,
                  min: 0,
                  max: 360,
                  activeColor: widget.settings.clockColor,
                  onChanged: (val) {
                    setState(() => _hue = val);
                    _updateCustomColor();
                  },
                ),

                Text('Saturation: ${(_saturation * 100).toInt()}%', style: const TextStyle(color: Colors.white70, fontSize: 12)),
                Slider(
                  value: _saturation,
                  min: 0,
                  max: 1,
                  activeColor: widget.settings.clockColor,
                  onChanged: (val) {
                    setState(() => _saturation = val);
                    _updateCustomColor();
                  },
                ),

                Text('Brightness: ${(_value * 100).toInt()}%', style: const TextStyle(color: Colors.white70, fontSize: 12)),
                Slider(
                  value: _value,
                  min: 0.2,
                  max: 1,
                  activeColor: widget.settings.clockColor,
                  onChanged: (val) {
                    setState(() => _value = val);
                    _updateCustomColor();
                  },
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

File: app/src/main/java/com/example/ui/screens/ColorScreen.kt
package com.example.ui.screens

import android.annotation.SuppressLint
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.data.ClockSettings
import com.example.ui.components.ClockDisplay
import java.time.LocalDateTime

/**
 * Preset clock colors required by the prompt
 */
val PRESET_CLOCK_COLORS = listOf(
    0xFFFFFFFF to "White",
    0xFF404040 to "Black / Dark",
    0xFFFF3B30 to "Red",
    0xFF007AFF to "Blue",
    0xFF34C759 to "Green",
    0xFFFFCC00 to "Yellow",
    0xFFAF52DE to "Purple",
    0xFFFF9500 to "Orange",
    0xFF00E5FF to "Cyan",
    0xFF39FF14 to "Neon Green",
    0xFFFF2D55 to "Hot Pink",
    0xFFFFB300 to "Amber"
)

/**
 * ColorScreen
 *
 * Educational screen providing:
 * 1. Live preview card of current time in selected color
 * 2. Preset color swatches (White, Black, Red, Blue, Green, Yellow, Purple, Orange, Cyan, etc.)
 * 3. Interactive Custom Color Picker (Hue, Saturation, Value sliders)
 *
 * Educational Note:
 * - Color in mobile UI is commonly represented as 32-bit ARGB (Alpha, Red, Green, Blue).
 * - For user-friendly custom color picking, HSV (Hue 0-360°, Saturation 0-1, Value 0-1) is
 *   significantly more intuitive than raw RGB because Hue corresponds directly to the color wheel.
 */
@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@SuppressLint("NewApi")
@Composable
fun ColorScreen(
    dateTime: LocalDateTime,
    settings: ClockSettings,
    onColorSelected: (Long) -> Unit,
    onBack: () -> Unit,
    modifier: Modifier = Modifier
) {
    // Sliders state for custom HSV picker
    var hue by remember { mutableFloatStateOf(180f) }
    var saturation by remember { mutableFloatStateOf(1.0f) }
    var value by remember { mutableFloatStateOf(1.0f) }

    val customColor = remember(hue, saturation, value) {
        val hsv = floatArrayOf(hue, saturation, value)
        val argb = android.graphics.Color.HSVToColor(hsv)
        Color(argb)
    }

    Scaffold(
        modifier = modifier.fillMaxSize(),
        containerColor = Color(0xFF0D1117),
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = "Clock Color",
                        color = Color.White,
                        fontWeight = FontWeight.Bold
                    )
                },
                navigationIcon = {
                    IconButton(
                        onClick = onBack,
                        modifier = Modifier.testTag("color_back_btn")
                    ) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Back",
                            tint = Color.White
                        )
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = Color(0xFF0D1117)
                )
            )
        }
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
                .padding(horizontal = 16.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(18.dp)
        ) {
            // Live Preview Card
            Card(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(20.dp),
                colors = CardDefaults.cardColors(
                    containerColor = Color.Black.copy(alpha = 0.4f)
                )
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 24.dp, horizontal = 12.dp),
                    contentAlignment = Alignment.Center
                ) {
                    ClockDisplay(
                        dateTime = dateTime,
                        settings = settings,
                        isFullScreen = false,
                        previewScale = 0.85f
                    )
                }
            }

            // Section 1: Presets
            Text(
                text = "Preset Colors",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
                color = Color.White
            )

            FlowRow(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                PRESET_CLOCK_COLORS.forEach { (colorValue, name) ->
                    val isSelected = settings.colorHex == colorValue
                    val swatchColor = Color(colorValue)

                    Surface(
                        onClick = { onColorSelected(colorValue) },
                        modifier = Modifier
                            .size(52.dp)
                            .testTag("preset_color_$name"),
                        shape = CircleShape,
                        color = swatchColor,
                        border = if (isSelected) {
                            androidx.compose.foundation.BorderStroke(3.dp, Color.White)
                        } else {
                            androidx.compose.foundation.BorderStroke(1.dp, Color.White.copy(alpha = 0.2f))
                        }
                    ) {
                        if (isSelected) {
                            Box(contentAlignment = Alignment.Center) {
                                Icon(
                                    imageVector = Icons.Default.Check,
                                    contentDescription = "Selected",
                                    tint = if (isLightColor(colorValue)) Color.Black else Color.White,
                                    modifier = Modifier.size(24.dp)
                                )
                            }
                        }
                    }
                }
            }

            // Section 2: Custom Color Picker (HSV)
            Card(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(20.dp),
                colors = CardDefaults.cardColors(
                    containerColor = Color.White.copy(alpha = 0.05f)
                )
            ) {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(18.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(
                            text = "Custom Color Picker",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color = Color.White
                        )

                        // Swatch & Hex preview
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Box(
                                modifier = Modifier
                                    .size(28.dp)
                                    .clip(CircleShape)
                                    .background(customColor)
                                    .border(1.dp, Color.White.copy(alpha = 0.4f), CircleShape)
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(
                                text = String.format("#%06X", (0xFFFFFF and customColor.toArgb())),
                                style = MaterialTheme.typography.labelMedium,
                                fontFamily = FontFamily.Monospace,
                                color = Color.White
                            )
                        }
                    }

                    // Hue Slider
                    Text(
                        text = "Hue (${hue.toInt()}°)",
                        style = MaterialTheme.typography.bodySmall,
                        color = Color.White.copy(alpha = 0.7f)
                    )
                    Slider(
                        value = hue,
                        onValueChange = {
                            hue = it
                            val hsv = floatArrayOf(hue, saturation, value)
                            val argb = android.graphics.Color.HSVToColor(hsv)
                            onColorSelected((argb.toLong() and 0xFFFFFFFFL))
                        },
                        valueRange = 0f..360f,
                        colors = SliderDefaults.colors(
                            thumbColor = customColor,
                            activeTrackColor = customColor
                        )
                    )

                    // Saturation Slider
                    Text(
                        text = "Saturation (${(saturation * 100).toInt()}%)",
                        style = MaterialTheme.typography.bodySmall,
                        color = Color.White.copy(alpha = 0.7f)
                    )
                    Slider(
                        value = saturation,
                        onValueChange = {
                            saturation = it
                            val hsv = floatArrayOf(hue, saturation, value)
                            val argb = android.graphics.Color.HSVToColor(hsv)
                            onColorSelected((argb.toLong() and 0xFFFFFFFFL))
                        },
                        valueRange = 0f..1f,
                        colors = SliderDefaults.colors(
                            thumbColor = customColor,
                            activeTrackColor = customColor
                        )
                    )

                    // Brightness (Value) Slider
                    Text(
                        text = "Brightness (${(value * 100).toInt()}%)",
                        style = MaterialTheme.typography.bodySmall,
                        color = Color.White.copy(alpha = 0.7f)
                    )
                    Slider(
                        value = value,
                        onValueChange = {
                            value = it
                            val hsv = floatArrayOf(hue, saturation, value)
                            val argb = android.graphics.Color.HSVToColor(hsv)
                            onColorSelected((argb.toLong() and 0xFFFFFFFFL))
                        },
                        valueRange = 0.2f..1f,
                        colors = SliderDefaults.colors(
                            thumbColor = customColor,
                            activeTrackColor = customColor
                        )
                    )
                }
            }

            Spacer(modifier = Modifier.height(20.dp))
        }
    }
}
